### PR TITLE
EVG-18537: implement parser project S3 storage

### PR DESCRIPTION
**Pull request by Kimchelly**
Jira: https://jira.mongodb.org/browse/EVG-18537

### Description 
* Implement parser project storage in S3. It's not used anywhere for now, there will be a follow-up to actually use it.
* Add some more docs about the `model.ParserProject` and how it's different from the `model.Project`.

### Testing
* Add integration tests for storing/retrieving parser projects from S3.
* Tested in staging by making a patch that kept its parser project in S3, then verified that generate.tasks worked with the S3 parser project.

### CODE REVIEW

File: model/project_parser_s3.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	if err != nil {
+		return nil, errors.Wrapf(err, "reading parser project '%s'", id)
+	}
+
+	var pp ParserProject
+	if err := bson.Unmarshal(b, &pp); err != nil {
+		return nil, errors.Wrapf(err, "unmarshalling parser project '%s' from BSON", id)
+	}
+
+	// TODO: this is in the equivalent DB method, but is it necessary?
```
Comment (Kimchelly) [inline model/project_parser_s3.go:78]: Not sure what this is for, but findOne for the parser project from the DB does this, so I just copied the logic.

File: model/project_parser_s3.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	return &pp, nil
+}
+
+// FindOneByIDWithFields finds a parser project using its ID from S3 and returns
+// the parser project with only the requested fields populated. This is not any
+// more efficient than FindOneByID. If the context errors, it will return the
+// context error.
+func (s *ParserProjectS3Storage) FindOneByIDWithFields(ctx context.Context, id string, fields ...string) (*ParserProject, error) {
```
Comment (bynn) [inline model/project_parser_s3.go:90]: Are we supposed to be doing something with fields?

File: model/project_parser_s3.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+// FindOneByID finds a parser project in S3 using its ID. If the context errors,
+// it will return the context error.
+func (s *ParserProjectS3Storage) FindOneByID(ctx context.Context, id string) (*ParserProject, error) {
+	if s.closed {
+		return nil, errors.New("cannot access parser project S3 storage when it is closed")
+	}
+
+	r, err := s.bucket.Get(ctx, id)
+	if pail.IsKeyNotFoundError(err) {
```
Comment (bynn) [inline model/project_parser_s3.go:60]: Why don't we want to error if we don't find anything?

**Review by bynn — CHANGES_REQUESTED** [review]

File: model/project_parser_s3.go (hunk 2 above)
Comment (Kimchelly) [reply model/project_parser_s3.go:90]: It's intentionally ignored. This method only exists to make it compatible with [the DB equivalent](https://github.com/evergreen-ci/evergreen/blob/23e6c4bb142bc3273995025021c4b285bd084e60/model/project_parser_db.go#L134-L136).

File: model/project_parser_s3.go (hunk 3 above)
Comment (Kimchelly) [reply model/project_parser_s3.go:60]: It's to follow the behavior of [the existing implementation for the DB](https://github.com/evergreen-ci/evergreen/blob/23e6c4bb142bc3273995025021c4b285bd084e60/model/project_parser_db.go#L71-L73).

**Pull request merged**